text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,659 @@
+
+### Span names, kinds, and attributes
+
+#### Span name
+
+The span name should be descriptive and make it clear, what operation a span
+describes. In the context of messaging systems, this means that a span should
+at the very least make clear that it refers to a messaging system, in addition
+it needs to make clear what particular [messaging operation](#operation-name)
+it refers to.
+
+Ideally, the span name also contains the destination name of the messages it
+refers to. However, a destination name should only be added to the span name
+when it is of low cardinality. This is usually the case when the destination
+name is a meaningful and manually configured name (like a manually configure
+queue or topic name), it is usually not the case if the destination name is an
+auto-generated identifier (like a conversation id or an auto-generated name for
+an anonymous destination).
+
+> The span name SHOULD consist of the name of the messaging system followed by
+> an [operation name](#operation-name). The destination name MAY be appended if
+> it is of low cardinality.
+
+##### Examples
+
+* `kafka publish shop.orders`
+* `rabbitmq receive print_jobs`
+* `AmazonSQS deliver`
+* `activemq settle`
+
+#### Operation name
+
+The following operations related to messages are covered by these semantic
+conventions:
+
+| Operation name | Description |
+|----------------|-------------|
+| `publish`      | One ore more messages are provided for publishing to an intermediary. |
+| `create`       | A message is created. |
+| `receive`      | One or more messages are requested by a consumer. |
+| `deliver`      | One or more message are passed to a consumer. |
+| `settle`       | One or more message are settled. |
+
+For further details about each of those operations refer, to the [section about trace structure](#trace-structure).
+
+#### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. `INTERNAL` otherwise. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+| `settle`       | `INTERNAL` |
+
+Setting span kinds according to this table ensures, that span links between
+consumer and producers always go from a `PRODUCER` span on the producer side to
+a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
+linked traces without the need of additional semantic hints.
+
+#### Attributes
+
+Attribute             | Type   | Requirement Level
+----------------------|--------|---------
+[`messaging.system`](#messagingsystem) | string | Yes
+[`messaging.operation`](#messagingoperation) | string | Required
+[`messaging.destination.name`](#messagingdestinationname) | string | For producer spans
```

Comment:
Done some research on `messaging.destination.name`:
- [kafka](https://stackoverflow.com/questions/39093488/can-single-kafka-producer-produce-messages-to-multiple-topics-and-how), [rocketmq](https://rocketmq.apache.org/docs/batch-example/) can publish a batch where different messages are sent to different topics. The only way out of it I see is through link per message with attributes. 
- Azure EventGrid either does not have a destination name at all and is fully identified by host or has topic-name per message and not per *send* call, i.e. same issue as above

Also, [Pulsar](https://github.com/streamnative/examples/blob/master/clients/pubsub/examples/delayed-message.md) that have tenant/namespace/topic structure, does not really care about it on the client API level. E.g, `public/dlt-example/dlt-example-topic` is passed as a string to the producer. I believe the whole thing should go to the `messaging.destination.name`

With this in mind, `messaging.destination.name` should be a conditionally required attribute (when available and when all messages in a batch are published to the same topic). 
And the contract is that it uniquely identifies topic/queue/subject/entity within either
- `net.peer.name:net.peer.port` (not available for GCP, AWS or vanilla JMS) or
- `cloud.account.id` + `cloud.region` for AWS - check out [AWS lambda otel samples](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/instrumentation/aws-lambda.md)
- `cloud.account.id` for GCP

It's probably hard to build UI with so few quarantees.

Reply:
For cloud messaging systems, I think we should go for the cloud-native resource ID, like the full ARN in AWS (see faas.id spec in current specification, it has some general wording that could be extracted / reused here).

Reply:
I wanted to see if we can take a step back with this. I was thinking about it and wanted to see your opinions on: 

- This varies a lot between different systems
- But it still seem like there is _always_ something that can be added here that uniquely identifies the destination

Given those facts, couldn't leave this still as `Required`? We then give _some_ hints on what instrumentation should usually put in here, but ultimately it's on them to know exactly how this should be composed. 

I feel we are running into circles with this attribute. Maybe we should make it more flexible and allow each individual messaging system SDK/Instrumentation do what's best for them. WDYT?

Reply:
This definitely needs a per-system definition, much like faas.id, db.name, etc.. But I don't think each instrumentation should decide on their own, at least not when it is likely that there will be different instrumentations for the same messaging system.